workflow/controller/steps.go

```diff
@@ -462,13 +462,34 @@ func (woc *wfOperationCtx) resolveReferences(stepGroup []wfv1.WorkflowStep, scop
 				if art.Optional {
 					continue
 				}
-				return nil, fmt.Errorf("unable to resolve references: %s", err)
+				return fmt.Errorf("unable to resolve references: %s", err)
 			}
 			resolvedArt.Name = art.Name
 			newStep.Arguments.Artifacts[j] = *resolvedArt
 		}
 
 		newStepGroup[i] = newStep
+		return nil
+	}
+
+	parallelStepNum := make(chan string, 500)
+	var resolveReferenceError error
+	var wg sync.WaitGroup
+	for i, step := range stepGroup {
+		parallelStepNum <- step.Name
+		wg.Add(1)
+		go func(i int, step wfv1.WorkflowStep) {
+			defer wg.Done()
+			if err := resolveStepReferences(i, step, newStepGroup); err != nil {
+				woc.log.WithFields(log.Fields{"stepName": step.Name}).WithError(err).Error(err.Error())
+				resolveReferenceError = err
+			}
+			<-parallelStepNum
```

Comment:
What's the purpose of `parallelStepNum` channel?

Reply:
> What's the purpose of `parallelStepNum` channel?

Just control the number of concurrencies， like “parallelPodNum” in workflow/controller/operator.go

Reply:
oh, neat. Would you mind adding a comment that that's what you're doing?

Reply:
Okay,I will add some comments.

Reply:
thank you!